Sources/SourceKitLSP/DocumentManager.swift

```diff
@@ -115,17 +152,56 @@ public final class DocumentManager {
         }
 
         if let range = edit.range  {
-
           document.latestLineTable.replace(
             fromLine: range.lowerBound.line,
             utf16Offset: range.lowerBound.utf16index,
             toLine: range.upperBound.line,
             utf16Offset: range.upperBound.utf16index,
             with: edit.text)
+          
+          // Remove all tokens in the updated range and shift later ones.
+
+          let previousLineCount = 1 + range.upperBound.line - range.lowerBound.line
+          let newLines = edit.text.split(separator: "\n", omittingEmptySubsequences: false)
+          let lastLineReplaceLength = (
+            range.lowerBound.line == range.upperBound.line ? range.upperBound.utf16index : 0
+          ) - range.lowerBound.utf16index
+          let lastLineLengthDelta = newLines.last!.count - lastLineReplaceLength
+          let lineDelta = newLines.count - previousLineCount
 
+          func isTokenBounding(character: Character) -> Bool {
```

Comment:
E.g. changing `let` to `alet` should remove the token, ` let` should not.

I agree, the logic here is a bit ad-hoc, I'm not sure what the optimal solution is though. Ideally, sourcekitd could tell us which tokens get removed from the syntaxmap after an edit, that way we wouldn't have to deal with edge cases like multi-line comments, where tokens outside the edit range have to be removed automatically.

Reply:
I just looked up how the syntax map is used and AFAICT there shouldn’t be any special handling regarding whitespaces necessary. All tokens whose start position is `>=` end position of the edit should be kept.

I didn’t have time to look at it in too much detail, but `key.offset` and `key.length` from the SourceKit response to the edit might be exactly the range for which a new syntax map is being provided. Like [here](https://github.com/apple/swift/blob/ad6a6e46561427992bc88317816ad2ea5bd14990/test/SourceKit/SyntaxMapData/syntaxmap-edit-block-comment.swift#L46-L47) where we get a `key.length > 1` for an edit of length 1.

Reply:
Yes, but that wouldn't solve the example I gave, if I understand you correctly. `let` is a token whose start position is == the end of the edit range (thus also >=) that wouldn't be removed if the user changes it into e.g. `alet`.

Also, what happens e.g. if the user types
```
[1. edit: /*]
let x = 4
[2. edit: */]
```

Since the `let` token overlaps with neither edit, how would sourcekit-lsp know that `let` should be removed?

Reply:
The `offset` and `length` returned by `sourcekitd` should include the `let` token. And I think the same goes for comments. So, you would need to do all the checks not based on the length and offset you pass in the request, but based on the ones you get back in the response, if that makes sense.

For example, when I change `let x = 1` to `alet = 1` I get the following response with offset 0 and length 4:
```bash
$ cat ~/Desktop/test.swift
let x = 1
$ ./sourcekitd-test -req=open /Users/alex/Desktop/test.swift -- /Users/alex/Desktop/test.swift == -req=edit -offset=0 -length=0 -replace="a" /Users/alex/Desktop/test.swift -print-raw-response -- /Users/alex/Desktop/test.swift
```
<details>
<summary>Response</summary>

```
{
  key.request: source.request.editor.open,
  key.name: "/Users/alex/Desktop/test.swift",
  key.compilerargs: [
    "/Users/alex/Desktop/test.swift"
  ],
  key.sourcefile: "/Users/alex/Desktop/test.swift"
}
{
  key.request: source.request.editor.replacetext,
  key.name: "/Users/alex/Desktop/test.swift",
  key.compilerargs: [
    "/Users/alex/Desktop/test.swift"
  ],
  key.offset: 0,
  key.length: 0,
  key.sourcefile: "/Users/alex/Desktop/test.swift",
  key.sourcetext: "a"
}
{
  key.offset: 0, // <---- These are the important information SourceKit is passing back to us
  key.length: 4, // <---- These are the important information SourceKit is passing back to us
  key.diagnostic_stage: source.diagnostic.stage.swift.parse,
  key.syntaxmap: [
    {
      key.kind: source.lang.swift.syntaxtype.identifier,
      key.offset: 0,
      key.length: 4
    }
  ],
  key.substructure: [
  ],
  key.diagnostics: [
    {
      key.line: 1,
      key.column: 5,
      key.filepath: "/Users/alex/Desktop/test.swift",
      key.severity: source.diagnostic.severity.error,
      key.id: "statement_same_line_without_semi",
      key.description: "consecutive statements on a line must be separated by ';'",
      key.diagnostic_stage: source.diagnostic.stage.swift.parse,
      key.fixits: [
        {
          key.offset: 4,
          key.length: 0,
          key.sourcetext: ";"
        }
      ]
    }
  ]
}
```
</details>